golos.publication/golos.publication.cpp

```diff
@@ -435,41 +440,88 @@ void publication::close_message(structures::mssgid message_id) {
             eosio_assert(state.funds.amount >= 0, "LOGIC ERROR! publication::payrewards: state.funds < 0");
         }
 
-        auto new_rshares = WP(state.rshares) - wdfp_t(FP(mssg_itr->state.netshares));
-        auto new_rsharesfn = WP(state.rsharesfn) - wdfp_t(sharesfn);
+        auto rshares = wdfp_t(FP(mssg_itr->state.netshares));
+        auto new_rshares = WP(state.rshares) - rshares;
+
+        auto rsharesfn = wdfp_t(sharesfn);
+        auto new_rsharesfn = WP(state.rsharesfn) - rsharesfn;
 
         eosio_assert(new_rshares >= 0, "LOGIC ERROR! publication::payrewards: new_rshares < 0");
         eosio_assert(new_rsharesfn >= 0, "LOGIC ERROR! publication::payrewards: new_rsharesfn < 0");
 
         state.rshares = new_rshares.data();
         state.rsharesfn = new_rsharesfn.data();
+        
+        mssg_reward = asset(payout, state.funds.symbol); 
     }
 
-    elaf_t percent(elai_t(mssg_itr->curators_prcnt) / elai_t(config::_100percent));
-    auto curation_payout = int_cast(percent * elai_t(payout));
+    message_table.modify(mssg_itr, name(), [&]( auto &item ) {
+            item.closed = true;
+            item.mssg_reward = mssg_reward;
+        });
 
-    eosio_assert((curation_payout <= payout) && (curation_payout >= 0), "publication::payrewards: wrong curation_payout");
-    auto unclaimed_rewards = pay_curators(message_id.author, mssg_itr->id, curation_payout, FP(mssg_itr->state.sumcuratorsw), state.funds.symbol, get_memo("curators", message_id));
+    bool pool_erased = false;
+    state.msgs--;
+    if(state.msgs == 0) {
+        if(pool != --pools.end()) {
+            send_poolerase_event(*pool);
 
-    eosio_assert(unclaimed_rewards >= 0, "publication::payrewards: unclaimed_rewards < 0");
+            pools.erase(pool);
+            pool_erased = true;
+        }
+    }
+    if(!pool_erased)
+        pools.modify(pool, _self, [&](auto &item) {
+            item.state = state;
+            send_poolstate_event(item);
+        });
 
-    state.funds.amount += unclaimed_rewards;
-    payout -= curation_payout;
+    transaction trx;
+    trx.actions.emplace_back(action{permission_level(_self, config::active_name), _self, "paymssgrwrd"_n, std::make_tuple(mssg_reward, message_id)});
+    trx.delay_sec = 0;
+    trx.send((static_cast<uint128_t>(mssg_itr->id) << 64) | message_id.author.value, _self);
+}
 
+void publication::paymssgrwrd(asset payout, structures::mssgid message_id) {
```

Comment:
Don't pass `payout` as argument to `paymssgrwrd` action. In future `paymssgrwrd` may be used by Golos.io for technical reason (if `paymssgrwrd` fails due expiration time or some another reason). If in this case actor calls this action with large payout value, misuse of token in reward pool will happen.
Please remove `payout` from arguments and use `mssg_reward` field.